src/main/java/com/gregtechceu/gtceu/api/data/worldgen/generator/veins/CuboidVeinGenerator.java

```diff
@@ -0,0 +1,285 @@
+
+    public static final Codec<CuboidVeinGenerator> CODEC = RecordCodecBuilder.create(instance -> instance.group(
+        LAYER_CODEC.fieldOf("top").forGetter(val -> val.top),
+        LAYER_CODEC.fieldOf("middle").forGetter(val -> val.middle),
+        LAYER_CODEC.fieldOf("bottom").forGetter(val -> val.bottom),
+        LAYER_CODEC.fieldOf("spread").forGetter(val -> val.spread),
+        Codec.INT.fieldOf("min_y").forGetter(val -> val.minY),
+        Codec.INT.fieldOf("max_y").forGetter(val -> val.maxY)
+    ).apply(instance, CuboidVeinGenerator::new));
+
+    private Either<List<OreConfiguration.TargetBlockState>, Material> top;
+    private Either<List<OreConfiguration.TargetBlockState>, Material> middle;
+    private Either<List<OreConfiguration.TargetBlockState>, Material> bottom;
+    private Either<List<OreConfiguration.TargetBlockState>, Material> spread;
+
+    private int minY, maxY;
+
+    public CuboidVeinGenerator(GTOreDefinition entry) {
+        super(entry);
+    }
+
+    @Override
+    public List<Map.Entry<Either<BlockState, Material>, Integer>> getAllEntries() {
+        List<Map.Entry<Either<BlockState, Material>, Integer>> result = new ArrayList<>();
+        // Entries' values are counted based on how many layers the entry is in.
+        top.map(blockStates -> blockStates.stream().map(state -> Either.<BlockState, Material>left(state.state)),
+                material -> Stream.of(Either.<BlockState, Material>right(material)))
+            .forEach(entry -> result.add(Map.entry(entry, 2)));
+        middle.map(blockStates -> blockStates.stream().map(state -> Either.<BlockState, Material>left(state.state)),
+                material -> Stream.of(Either.<BlockState, Material>right(material)))
+            .forEach(entry -> result.add(Map.entry(entry, 3)));
+        bottom.map(blockStates -> blockStates.stream().map(state -> Either.<BlockState, Material>left(state.state)),
+                material -> Stream.of(Either.<BlockState, Material>right(material)))
+            .forEach(entry -> result.add(Map.entry(entry, 4)));
+        spread.map(blockStates -> blockStates.stream().map(state -> Either.<BlockState, Material>left(state.state)),
+                material -> Stream.of(Either.<BlockState, Material>right(material)))
+            .forEach(entry -> result.add(Map.entry(entry, 7)));
+        return result;
+    }
+
+    @Override
+    public Map<BlockPos, OreBlockPlacer> generate(WorldGenLevel level, RandomSource random, GTOreDefinition entry, BlockPos origin, ChunkPos currentChunk) {
+        Map<BlockPos, OreBlockPlacer> generatedBlocks = new Object2ObjectOpenHashMap<>();
+
+        // determine density based on distance from the origin chunk
+        // this makes the vein more concentrated towards the center
+        double xLength = currentChunk.x - origin.getX();
+        double zLength = currentChunk.z - origin.getZ();
```

Comment:
The currently generating chunk's position shouldn't affect the generated vein in any way, since we can't know in advance which chunks are generated in which order.

Keep in mind that `generate()` is only called once and will be cached for subsequent nearby chunks to re-use, so this would effectively change the density for larger parts of the generated vein than what's intended.